File: chapter-12/exercise-12.2.cpp
/*
12.2 Reverse String

Implement a function void reverse(char* str) in C or C++ which reverses a null
terminated string.
*/

#include <string>

void reverse(char *str) {
    char* end = str;
    char tmp;
    if (str) {
        while (*end) {
            ++end;
        }
        --end;

        while (str < end) {
            tmp = *str;
            *str++ = *end
            *end-- = tmp;
        }
    }
}

File: chapter-12/exercise-12.2.py
"""
12.2 Reverse String

Implement a function to reverse a string
"""

from collections import deque

def reverse(string):
    n = len(string)
    stack = deque()

    for i in range(0, n, 1):
        stack.append((string[i]))

    string = ""
    for i in range(0, n, 1):
        string += stack.pop()

    return string